Overhaul addon documentation

Charrondev opened the pull request:
I've overhauled the addon documentation.

- Documentation of the addon format and `addon.json` config.
- Updating other docs to use use the new format.
- Update coding standards in addon docs.
- Unifying and updating of Theme and Plugin documentation.

Now that we have version tags in our documentation, I've separated the old PluginInfo/ThemeInfo docs from the addon.json docs and added the versioning tags.

DaazKu commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,278 @@
+
+#### key
+
+The theme key. This should be unique and ***must exactly match*** the folder name, including capitalization. So an addon with the key `example-addon` would need to be located in a `plugins/example-addon` directory in your Vanilla installation. Likewise a theme with the key `UpperCaseTheme` would need to be located in a `themes/UpperCaseTheme` directory inside your Vanilla installation.
```
Since it will be used internally we should specify that we want lower case folders. (that's the new standard)

DaazKu commented on content/developer/addons/addon-quickstart.md:
````diff
@@ -0,0 +1,233 @@
+
+```json
+"registerPermissions": {
+    "FancyAddon.DoStuff.Add" => 1
````
Should be `"FancyAddon.Stuff.Add": 1`

DaazKu commented on content/developer/addons/addon-quickstart.md:
````diff
@@ -0,0 +1,233 @@
+
+```json
+"registerPermissions": {
+    "Plugins.FancyAddon.PermissionName": "Garden.Settings.Manage"
````
Plugins.* is now deprecated. It should be `AddonName.Target.ImperativeVerb` (most of the time)

DaazKu commented on content/developer/addons/addon-quickstart.md:
````diff
@@ -0,0 +1,233 @@
+}
+```
+
+Use the `$sender` variable to call methods from the controller or model that triggered your code. 
````
$sender can an object of any class.

Generic docblock
```php
    /**
     * @param object $sender Sending object instance.
     * @param array $args Event's arguments.
     */
    public function base_someEvent_handler($sender, $args) {
```

Specific docblock
```php
    /**
     * @param CustomClass $sender Sending CustomClass instance.
     * @param array $args Event's arguments.
     */
    public function customClass_someEvent_handler($sender /* Could be named $customClass for clarity */, $args) {
```

DaazKu commented on content/developer/addons/addon-quickstart.md:
```diff
@@ -0,0 +1,233 @@
+   $this->commonMethod($sender, $args);
+}
+
+public function categoriesController_someOtherEvent_handler($sender, $args) {
```
Usually you will use a common method for the same event ;)

DaazKu commented on content/developer/addons/addon-quickstart.md:
```diff
@@ -0,0 +1,233 @@
+
+Note the second parameter is the folder.
+
+To include a CSS file in a specific page (e.g. the Discussions list):
```
This would target a specific controller not a specific page.
One controller can have multiple pages.

DaazKu commented on content/developer/addons/addon-quickstart.md:
````diff
@@ -0,0 +1,233 @@
+To include a CSS file in a specific page (e.g. the Discussions list):
+
+```php
+public function discussionsController_render_before($Sender) {
````
Typo: $Sender -> $sender

slafleche commented on content/developer/theming/about-php.md:
```diff
@@ -1,91 +0,0 @@
----
```
@Charrondev Don't delete the about.php page. If anybody has links to it, we don't want them to land on a 404 page. Mark it as deprecated, keeping relevant information and link to the new pages talking about the json format.

Charrondev replied:
I've kept the information in the add-on information page and added an alias to is. I've tested the redirect.

DaazKu replied:
Even better! Ask for a redirect from OPs

Charrondev replied:
The aliases field in the front-matter handles the redirects, so I've already put in redirects for any page I've moved or that are related to the old one.

Charrondev replied:
I've now made a separate PR for moving the smarty docs. It included redirects.

https://github.com/vanilla/docs/pull/146

slafleche commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,165 @@
+  developer:
+    weight: 10
+    parent: addons
+version:
```
The global versioning is not working because the readme is out of date. Use: "versioning" instead of "version". I'll update the readme. Issue: https://github.com/vanilla/docs/issues/147

Charrondev commented:
I've updated this PR with following:
- Content improvements suggested by Alex
- Fixed heading hierarchy in all affected pages (h2->h3->h4)
- Verified all necessary aliases are in place
- Fixed ordering and improved naming of pages in the sidebar
- Used versioning tags in the front-matter where appropriate.

slafleche commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,164 @@
+---
+title: Addon Information
+tags:
```
add addon-info, addon

Charrondev replied:
Added

slafleche commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,164 @@
+
+Every Vanilla addons must declare certain information about itself. This includes:
+
+- Addon Metadata
```
If these are subsections in this page, make anchor links

Charrondev replied:
They are not.

slafleche commented on content/developer/addons/addon-visibility.md:
```diff
@@ -0,0 +1,100 @@
+    weight: 11
+    parent: addons
+aliases:
+- /theming/visibility
```
Was that a real link? Is this a remnant from an old file?

Charrondev replied:
It seems you added it when the file was creating. It's probably not necessary.

https://github.com/vanilla/docs/blame/2e283f7466eda728db98ebe6c2a0da4c30915d73/content/developer/theming/themevisibility.md#L11

slafleche commented on content/developer/embedding/index.md:
```diff
@@ -7,7 +7,6 @@ category: help
     identifier: embedding
-    weight: 100
 aliases:
 - /features/embedding
```
Incorrect Alias

Charrondev replied:
It seems this file is actually in 2 places. I'm going to remove the one in the help section and create an alias to it in the one on the developer side.

Charrondev replied:
I've implemented this feedback

slafleche commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,164 @@
+            "email": "adam.c@vanillaforums.com"
+        }
+    ],
+    "layout": {
```
I would add the a sample "sites" node, because it comes up frequently as a problem with clients.

Charrondev replied:
I've added it on both

slafleche commented on content/developer/addons/addon-info.md:
```diff
@@ -0,0 +1,164 @@
+
+### Details
+
+#### key
```
It would be nice to add sample examples for each of these

slafleche commented on content/developer/addons/theme-quickstart.md:
```diff
@@ -0,0 +1,135 @@
+
+### Visibility
+
+If you've got [Vanilla Cloud](http://vanillaforums.com), make sure to also [set your themes's visibility](/developer/adddons/addon-visibility).
```
Broken link

slafleche commented on content/developer/addons/theming-with-css.md:
````diff
@@ -1,52 +1,44 @@
-]
-```
-The layout options for discussions are currently 'table' or 'modern'. For categories, you can choose 'table', 'modern', or 'mixed'. The mixed layout includes discussions on the categories page.
+Vanilla offers either a HTML table layout or list layout for its Categories and Discussions listings. Each layout has its own set of CSS classes, so it's best to decide which layout you'd like to use before you start styling your theme. This layout can be selected from the dashboard's Homepage section. You can also specify your theme's layout in the $ThemeInfo array in your theme's [addon.json](/ ...
````
Broken link to Theming Quickstart Guide

slafleche commented on content/developer/addons/view-overrides.md:
```diff
@@ -112,7 +115,7 @@ To see how it all fits together, here's a default master view using Smarty:
-### Overriding other views
+## Overriding other views
 
 Now that you know how to override and configure the master view, you may want to configure the content of the assets and modules. Before we get there though, a warning: overriding a view can be a rather severe addition to a theme. Once a view is overridden, it diverges from core Vanilla. As such, your new view may not always be supported in future versions of Vanilla, which may cause problems dow ...
```
Broken link

slafleche replied:
This link appears more than once.

slafleche commented on content/developer/addons/index.md:
```diff
@@ -0,0 +1,60 @@
+
+Vanilla is a very customizable, flexible platform because there are many ways to hook into it without ever modifying its code. There are normal vanilla [addons](/developer/addons/addon-quickstart) and [themes](/developer/addons/theme-quickstart). Themes and addons are very similar, but serve different purposes.
+
+## Addons and Plugins and Themes - Oh My!
```
Since we're deleting the old plugins pages, it might be good to have a little blurb on pages being redirected about the change...

slafleche commented on content/developer/theming/index.md:
```diff
@@ -1,26 +0,0 @@
----
-title: Theming
```
Needs an alias to this page

Charrondev replied:
Added

Charrondev commented:
I've pushed some changes to this branch.

- Much improved definitions of what a theme/plugin/addon is and their differences
    - Cross-links to this definition in the beginning of most other addon pages
- Additional tags on the addon pages
- Fixed broken links
- Removed a duplicated page and added an alias to it.
- Removed `/theming/something` aliases and added additional `/developer/theming` aliases
